src/Map.ts

```diff
@@ -0,0 +1,825 @@
+import { sort } from './Array'
+import { Applicative, Applicative1, Applicative2, Applicative2C, Applicative3, Applicative3C } from './Applicative'
+import { Compactable2, Separated } from './Compactable'
+import { Either } from './Either'
+import { FilterableWithIndex2C } from './FilterableWithIndex'
+import { Foldable, Foldable1, Foldable2, Foldable3 } from './Foldable'
+import { Foldable2v2C } from './Foldable2v'
+import { FoldableWithIndex2C } from './FoldableWithIndex'
+import { Predicate, Refinement, phantom, tuple } from './function'
+import { HKT, Type, Type2, Type3, URIS, URIS2, URIS3 } from './HKT'
+import { Monoid } from './Monoid'
+import { Option, none, some } from './Option'
+import { Ord } from './Ord'
+import { Setoid, fromEquals } from './Setoid'
+import { TraversableWithIndex2C } from './TraversableWithIndex'
+import { Unfoldable, Unfoldable1 } from './Unfoldable'
+import { Semigroup } from './Semigroup'
+import { Witherable2C } from './Witherable'
+import { FunctorWithIndex2C } from './FunctorWithIndex'
+import { Functor2 } from './Functor'
+import { Traversable2v2C } from './Traversable2v'
+import { Filterable2 } from './Filterable'
+
+declare module './HKT' {
+  interface URI2HKT2<L, A> {
+    Map: Map<L, A>
+  }
+}
+
+export const URI = 'Map'
+
+export type URI = typeof URI
+
+/**
+ * Calculate the number of key/value pairs in a Map
+ *
+ * @since 1.14.0
+ */
+export const size = <K, A>(d: Map<K, A>): number => d.size
+
+/**
+ * Test whether or not a Map is empty
+ *
+ * @since 1.14.0
+ */
+export const isEmpty = <K, A>(d: Map<K, A>): boolean => d.size === 0
+
+/**
+ * Test whether or not a key exists in a map
+ *
+ * @since 1.14.0
+ */
+export const has = <K>(S: Setoid<K>): (<A>(k: K, m: Map<K, A>) => boolean) => {
+  const lookupS = lookup(S)
+  return (k, m) => lookupS(k, m).isSome()
+}
+
+/**
+ * Test whether or not a value is a member of a map
+ *
+ * @since 1.14.0
+ */
+export const isMember = <A>(S: Setoid<A>): (<K>(a: A, m: Map<K, A>) => boolean) => {
+  return (a, m) => {
+    const as = m.values()
+    let e: IteratorResult<A>
+    while (!(e = as.next()).done) {
+      if (S.equals(a, e.value)) {
+        return true
+      }
+    }
+    return false
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const keys = <K>(O: Ord<K>): (<A>(m: Map<K, A>) => Array<K>) => {
+  const sortO = sort(O)
+  return m => {
+    const ks = Array.from(m.keys())
+    return sortO(ks)
+  }
+}
+
+/**
+ * @since 1.14.0
+ */
+export const keysSet = <K, A>(m: Map<K, A>): Set<K> => new Set(m.keys())
+
+/**
+ * @since 1.14.0
+ */
+export const collect = <K>(O: Ord<K>): (<A, B>(m: Map<K, A>, f: (k: K, a: A) => B) => Array<B>) => {
+  const lookupWithKeyO = lookupWithKey(O)
+  const keysO = keys(O)
+  return <A, B>(m: Map<K, A>, f: (k: K, a: A) => B): Array<B> => {
+    const out: Array<B> = []
+    const ks = keysO(m)
+    for (let key of ks) {
+      const o = lookupWithKeyO(key, m)
```

Comment:
this looks O(n^2), maybe we can leverage `get`

```diff
-const o = lookupWithKeyO(key, m)
-if (o.isSome()) {
-  out.push(f(key, o.value[1]))
-}
+out.push(f(key, m.get(key)!))
```

Reply:
I made this change. Marking it as resolved.